Rust/src/main.rs

```diff
@@ -107,57 +102,56 @@ fn drop_disk(board: &Board, column:u32, color:i32) -> Board {
-                                    // when loss is certain, avoid forfeiting the match, by shifting scores by depth...
-                                    match bscore {
-                                        1000000 | -1000000 => bscore - depth*color,
-                                        _ => bscore
-                                    })
-                                .collect();
-                            scores
-                        }
-                    };
-                    let all_data: Vec<_> = best_scores.iter().zip(valid_moves).collect();
-                    if *debug && depth == MAX_DEPTH {
-                        for (column, score) in all_data.clone() {
-                            println!("Depth {}, placing on {}, Score:{}\n", depth, column, score);
-                        }
-                    }
-                    let best = if maximize_or_minimize {
-                        all_data.iter().max()
-                    } else {
-                        all_data.iter().min()
-                    };
-                    match best {
-                        None => (None, 0),
-                        Some(x) => {
-                            let &(&best_score, best_move) = x;
-                            (Some(best_move),best_score)
-                        }
-                    }
-                },
-                _ => (Some(killer_moves[0].0), killer_moves[0].1),
-            }
+    let valid_moves: Vec<_> = (0..(WIDTH as u32)).filter(|column| board[0][*column as usize] == 0).collect();
+    if valid_moves.is_empty() {
+        return (None, score_board(board))
+    }
+
+    let target_score = if maximize_or_minimize {
+        ORANGE_WINS
+    } else {
+        YELLOW_WINS
+    };
+
+    let moves_and_boards: Vec<_> = valid_moves
+        .iter()
+        .map(|column| (*column, drop_disk(board, *column, color)))
+        .collect();
+    let moves_and_scores: Vec<_> = moves_and_boards
+        .iter()
+        .map(|&(column,board)| (column, score_board(&board)))
+        .collect();
+
+    if let Some(killer_move) = moves_and_scores.iter().find(|& &(_,score)| score == target_score) {
+        return (Some(killer_move.0), killer_move.1);
+    }
+
+    let best_scores: Vec<_> = match depth {
+        1 => moves_and_scores.iter().map(|x| x.1).collect(),
+        _ => moves_and_boards.iter().map(|x| &x.1)
+            .map(|&le_board| ab_minimax(!maximize_or_minimize, other_color(color), depth-1, &le_board, debug))
+            .map(|(_,bscore)| match bscore {
+                     // when loss is certain, avoid forfeiting the match, by shifting scores by depth...
+                     1000000 | -1000000 => bscore - depth*color,
+                     _ => bscore
+            })
+            .collect()
+    };
+    let all_data: Vec<_> = best_scores.iter().zip(valid_moves).collect();
+    if *debug && depth == MAX_DEPTH {
+        for &(column, score) in &all_data {
```

Comment:
Hi Steve!

I will study everything you sent in the coming weekend - and once I
understand what you did, merge them in :-)

Thanks, much appreciated!

On Mon, Nov 23, 2015 at 10:44 PM, Steven Allen <notifications@github.com>
wrote:

> In Rust/src/main.rs
> <https://github.com/ttsiodras/Score4/pull/2#discussion_r45657916>:
>
> > +    }
> > +
> > +    let best_scores: Vec<_> = match depth {
> > +        1 => moves_and_scores.iter().map(|x| x.1).collect(),
> > +        _ => moves_and_boards.iter().map(|x| &x.1)
> > +            .map(|&le_board| ab_minimax(!maximize_or_minimize, other_color(color), depth-1, &le_board, debug))
> > +            .map(|(_,bscore)| match bscore {
> > +                     // when loss is certain, avoid forfeiting the match, by shifting scores by depth...
> > +                     1000000 | -1000000 => bscore - depth*color,
> > +                     _ => bscore
> > +            })
> > +            .collect()
> > +    };
> > +    let all_data: Vec<_> = best_scores.iter().zip(valid_moves).collect();
> > +    if *debug && depth == MAX_DEPTH {
> > +        for &(column, score) in &all_data {
>
> IntoIterator<&T> is implemented on &Vec<T>. The pattern on the left hand
> side (&(column, score)) dereferences and destructures &T.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ttsiodras/Score4/pull/2/files#r45657916>.
>



-- 
e-mail: ttsiodras@gmail.com
web: http://users.softlab.ntua.gr/~ttsiod/index.html